File: rb/algorithms/lib/sorting/merge.rb
module Sorting
  class Merge
    def merge!(collection, lo, mid, hi)
      left, right = lo, (mid + 1)
      aux = collection.dup

      (lo..hi).each do |k|
        if left > mid                # left exhausted
          collection[k] = aux[right]
          right = right.next
        elsif right > hi             # right exhausted
          collection[k] = aux[left]
          left = left.next
        elsif aux[right] < aux[left]
          collection[k] = aux[right]
          right = right.next
        else
          collection[k] = aux[left]
          left = left.next
        end
      end
    end

    def sort!(collection)
      do_sort!(collection, 0, (collection.length - 1))
    end

    private

    def do_sort!(collection, lo, hi)
      return if hi <= lo
      mid = Integer(lo + ((hi - lo) / 2))
      do_sort!(collection, lo, mid)
      do_sort!(collection, (mid + 1), hi)
      merge!(collection, lo, mid, hi)
    end
  end
end

File: lib/sorting/merge.js
function merge(ary, lo, mid, hi) {
  var left = lo;
  var right = (mid + 1);
  var aux = ary.slice();

  for (var k = lo; k <= hi; k++) {
    if (left > mid) {         // left exhausted
      ary[k] = aux[right++];
    } else if (right > hi) { // right exhausted
      ary[k] = aux[left++];
    } else if (aux[right] < aux[left]) {
      ary[k] = aux[right++];
    } else {
      ary[k] = aux[left++];
    }
  }
};

function do_sort(ary, lo, hi) {
  if (hi <= lo) return;
  var mid = Math.floor(lo + ((hi - lo) / 2));
  do_sort(ary, lo, mid);
  do_sort(ary, (mid + 1), hi);
  merge(ary, lo, mid, hi);
}

module.exports = function (ary) {
  do_sort(ary, 0, (ary.length - 1));
  return ary;
};
